evm_arithmetization/src/generation/mod.rs

```diff
@@ -309,39 +312,220 @@ pub fn generate_traces<F: RichField + Extendable<D>, const D: usize>(
+    pub(crate) fn checkpoint(&mut self) -> GenerationStateCheckpoint {
+        match self {
+            Self::Generation(state) => state.checkpoint(),
+            Self::Interpreter(interpreter) => interpreter.checkpoint(),
+        }
+    }
+
+    /// Increments the `gas_used` register by a value `n`.
+    pub(crate) fn incr_gas(&mut self, n: u64) {
+        match self {
+            Self::Generation(state) => state.registers.gas_used += n,
+            Self::Interpreter(interpreter) => interpreter.generation_state.registers.gas_used += n,
+        }
+    }
+
+    /// Increments the `program_counter` register by a value `n`.
+    pub(crate) fn incr_pc(&mut self, n: usize) {
+        match self {
+            Self::Generation(state) => state.registers.program_counter += n,
+            Self::Interpreter(interpreter) => {
+                interpreter.generation_state.registers.program_counter += n
+            }
+        }
+    }
+
+    /// Returns a `State`'s registers.
+    pub(crate) fn get_registers(&self) -> RegistersState {
+        match self {
+            Self::Generation(state) => state.registers,
+            Self::Interpreter(interpreter) => interpreter.generation_state.registers,
+        }
+    }
+
+    /// Returns a `State`'s mutable registers.
+    pub(crate) fn get_mut_registers(&mut self) -> &mut RegistersState {
+        match self {
+            Self::Generation(state) => &mut state.registers,
+            Self::Interpreter(interpreter) => &mut interpreter.generation_state.registers,
+        }
+    }
+
+    /// Returns the value stored at address `address` in a `State`.
+    pub(crate) fn get_from_memory(&mut self, address: MemoryAddress) -> U256 {
+        match self {
+            Self::Generation(state) => state.memory.get(address, false, &HashMap::default()),
+            Self::Interpreter(interpreter) => interpreter.generation_state.memory.get(
+                address,
+                true,
+                &interpreter.preinitialized_segments,
+            ),
+        }
+    }
+
+    /// Returns a mutable `GenerationState` from a `State`.
+    pub(crate) fn get_mut_generation_state(&mut self) -> &mut GenerationState<F> {
+        match self {
+            Self::Generation(state) => state,
+            Self::Interpreter(interpreter) => &mut interpreter.generation_state,
+        }
+    }
+
+    /// Returns true if a `State` is a `GenerationState` and false otherwise.
+    pub(crate) fn is_generation_state(&mut self) -> bool {
+        match self {
+            Self::Generation(state) => true,
+            Self::Interpreter(interpreter) => false,
+        }
+    }
+
+    /// Increments the clock of an `Interpreter`'s clock.
+    pub(crate) fn incr_interpreter_clock(&mut self) {
+        match self {
+            Self::Generation(state) => {}
+            Self::Interpreter(interpreter) => interpreter.clock += 1,
+        }
+    }
+
+    /// Returns the value of a `State`'s clock.
```

Comment:
Maybe not in this PR, but we should unify the clock someday. I think we should add a `clock` register for the prover too, but it's not a strong opinion either.